File: java_solutions/91.decode-ways.java
/*
 * @lc app=leetcode id=91 lang=java
 *
 * [91] Decode Ways
 */

class Solution {
    public int numDecodings(String s) {
        return s.startsWith("0") ? 0 : java.util.stream.IntStream.range(1, s.length()).boxed().reduce(new int[]{1, 1}, (dp, i) -> new int[]{dp[1], (s.charAt(i) != '0' ? dp[1] : 0) + (s.charAt(i - 1) == '1' || (s.charAt(i - 1) == '2' && s.charAt(i) <= '6') ? dp[0] : 0)}, (a, b) -> a)[1];
    }
}

File: kotlin_solutions/91.decode-ways.kt
/*
 * @lc app=leetcode id=91 lang=kotlin
 *
 * [91] Decode Ways
 */

class Solution { fun numDecodings(s: String): Int = if (s.isEmpty() || s[0] == '0') 0 else s.drop(1).foldIndexed(1 to 1) { i, (prev, cur), c -> (if (c == '0') 0 else cur).let { oneDigit -> (if (s.substring(i, i + 2).toInt() in 10..26) prev else 0).let { twoDigit -> if (oneDigit == 0 && twoDigit == 0) return 0 else cur to (oneDigit + twoDigit) } } }.second }
